File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def dfs(root):
            if root is None:
                return 0
        
            l = dfs(root.left)
            if l == -1:
                return -1
            
            r = dfs(root.right)
            if r == -1 or abs(l - r) > 1:
                return -1
            
            return max(l, r) + 1
        
        if dfs(root) == -1:
            return False

        return True
        

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool flag = false;
        auto f = [&](auto &self, TreeNode* node) {
            if(!node) return 0;
            int height = 1;
            int left = self(self, node->left);
            int right = self(self, node->right);
            if(abs(left - right) <= 1) return height + max(left, right);
            else flag = true;
            return 0;
        };
        f(f, root);
        return flag ? false : true;
    }
};